packages/app/src/pages/transaction/index.tsx

```diff
@@ -0,0 +1,330 @@
+    contractSource,
+    version,
+    nonce,
+  });
+};
+
+export const Transaction: React.FC = () => {
+  const location = useLocation();
+  const { firstIdentity: identity } = useWallet();
+  const [pendingTransaction, setPendingTransaction] = useState<TransactionPayload | undefined>();
+  const [txHash, setTxHash] = useState('');
+  const [loading, setLoading] = useState(true);
+  const [contractSrc, setContractSrc] = useState('');
+  const [balance, setBalance] = useState(0);
+  const [nonce, setNonce] = useState(0);
+  const [error, setError] = useState<string | null>(null);
+  const client = getRPCClient();
+
+  if (!identity) {
+    throw new Error('User must be logged in.');
+  }
+
+  const tabs: Tab[] = [
+    {
+      title: 'Inputs',
+      content: <TabContent json={getInputJSON(pendingTransaction, identity)} />,
+      key: 'inputs',
+    },
+    {
+      title: (
+        <>
+          View Source
+          {/* Add this icon when we can link to the explorer */}
+          {/* <ExternalIcon display="inline-block" width="9px" ml={1} /> */}
+        </>
+      ),
+      content: (
+        <Box whiteSpace="pre" overflow="scroll" color="gray" maxHeight="200px">
+          {contractSrc}
+        </Box>
+      ),
+      key: 'source',
+      hide: pendingTransaction?.txType === 'stx-transfer',
+    },
+  ];
+
+  const setupAccountInfo = async () => {
+    if (identity) {
+      const account = await client.fetchAccount(
+        addressToString(identity.getSTXAddress(AddressVersion.TestnetSingleSig))
+      );
+      setBalance(account.balance.toNumber());
+      setNonce(account.nonce);
+    }
+  };
+
+  const setupWithState = async (tx: TransactionPayload) => {
+    if (tx.txType === 'contract-call') {
+      const contractSource = await client.fetchContractSource({
+        contractName: tx.contractName,
+        contractAddress: tx.contractAddress,
+      });
+      if (contractSource) {
+        setContractSrc(contractSource);
+        setPendingTransaction(tx);
+      } else {
+        doTrack(TRANSACTION_SIGN_ERROR, {
+          txType: pendingTransaction?.txType,
+          appName: pendingTransaction?.appDetails?.name,
+          error: 'Contract not found',
+        });
+        setError(`Unable to find contract ${tx.contractAddress}.${tx.contractName}`);
+      }
+    } else if (tx.txType === 'contract-deploy') {
+      console.log(tx);
+      setContractSrc(tx.contractSource);
+      setPendingTransaction(tx);
+    } else if (tx.txType === 'stx-transfer') {
+      setPendingTransaction(tx);
+    }
+    doTrack(TRANSACTION_SIGN_START, {
+      txType: tx.txType,
+      appName: tx.appDetails?.name,
+    });
+  };
+
+  const decodeRequest = async () => {
+    const urlParams = new URLSearchParams(location.search);
+    const requestToken = urlParams.get('request');
+    if (requestToken && identity) {
+      const token = decodeToken(requestToken);
+      const reqState = (token.payload as unknown) as TransactionPayload;
```

Comment:
Well, I'm not sure of a better way to do it. `token.payload` is basically just JSON - it can be anything. So, it's not much different than the "normal" method of casting JSON to a type. It could result in a runtime error, but that would likely get caught by the generic error handling on this page.